README.md
[![npm version](https://badge.fury.io/js/cordova-plugin-video-editor.svg)](https://badge.fury.io/js/cordova-plugin-video-editor)

This is a cordova plugin to assist in several video editing tasks such as:

* Transcoding
* Trimming
* Creating thumbnails from a video file (now at a specific time in the video)
* Getting info on a video - width, height, orientation, duration, size, & bitrate.

After looking at an article on [How Vine Satisfied Its Need for Speed](http://www.technologyreview.com/view/510511/how-vine-satisfies-its-need-for-speed/), it was clear Cordova/Phonegap needed a way to modify videos to be faster for app's that need that speed.

This plugin will address those concerns, hopefully.

## Installation
```
cordova plugin add cordova-plugin-video-editor
```
`VideoEditor` and `VideoEditorOptions` will be available in the window after deviceready.

## Usage

### Transcode a video

```javascript
// parameters passed to transcodeVideo
VideoEditor.transcodeVideo(
    success, // success cb
    error, // error cb
    {
        fileUri: 'file-uri-here', // the path to the video on the device
        outputFileName: 'output-name', // the file name for the transcoded video
        outputFileType: VideoEditorOptions.OutputFileType.MPEG4, // android is always mp4
        optimizeForNetworkUse: VideoEditorOptions.OptimizeForNetworkUse.YES, // ios only
        saveToLibrary: true, // optional, defaults to true
        maintainAspectRatio: true, // optional (ios only), defaults to true
        width: 640, // optional, see note below on width and height
        height: 640, // optional, see notes below on width and height
        videoBitrate: 1000000, // optional, bitrate in bits, defaults to 9 megabit (9000000)
        fps: 30, // optional (android only), defaults to 30
        audioChannels: 2, // optional, number of audio channels, defaults: iOS - 2, Android - as is
        audioSampleRate: 44100, // optional (ios only), sample rate for the audio, defaults to 44100
        audioBitrate: 128000, // optional, audio bitrate for the video in bits,  defaults: iOS - 128000 (128 kilobits), Android - as is or 128000
        skipVideoTranscodingIfAVC: true, // optional (android only), skip any transcoding actions (conversion/resizing/etc..) if the input video is avc video, defaults to false
        progress: function(info) {} // info will be a number from 0 to 100
    }
);
```
#### A note on width and height used by transcodeVideo
I recommend setting `maintainAspectRatio` to true.  When this option is true you can provide any width/height and the height provided will be used to calculate the new width for the output video.  If you set `maintainAspectRatio` false there is a good chance you'll end up with videos that are stretched and/or distorted.  Here is the simplified formula used on iOS when `maintainAspectRatio` is true -
```objective-c
aspectRatio = videoWidth / videoHeight;
outputWidth = height * aspectRatio;
outputHeight = outputWidth / aspectRatio;
```

Android will always use the aspect ratio of the input video to calculate the scaled output width and height.  Setting `maintainAspectRatio` on android will make not make a difference.

If you don't provide width and height to `transcodeVideo` the output video will have the same dimensions as the input video.

#### transcodeVideo example -
```javascript
// options used with transcodeVideo function
// VideoEditorOptions is global, no need to declare it
var VideoEditorOptions = {
    OptimizeForNetworkUse: {
        NO: 0,
        YES: 1
    },
    OutputFileType: {
        M4V: 0,
        MPEG4: 1,
        M4A: 2,
        QUICK_TIME: 3
    }
};
```
```javascript
// this example uses the cordova media capture plugin
navigator.device.capture.captureVideo(
    videoCaptureSuccess,
    videoCaptureError,
    {
        limit: 1,
        duration: 20
    }
);

function videoCaptureSuccess(mediaFiles) {
    // Wrap this below in a ~100 ms timeout on Android if
    // you just recorded the video using the capture plugin.
    // For some reason it is not available immediately in the file system.

    var file = mediaFiles[0];
    var videoFileName = 'video-name-here'; // I suggest a uuid

    VideoEditor.transcodeVideo(
        videoTranscodeSuccess,
        videoTranscodeError,
        {
            fileUri: file.fullPath,
            outputFileName: videoFileName,
            outputFileType: VideoEditorOptions.OutputFileType.MPEG4,
            optimizeForNetworkUse: VideoEditorOptions.OptimizeForNetworkUse.YES,
            saveToLibrary: true,
            maintainAspectRatio: true,
            width: 640,
            height: 640,
            videoBitrate: 1000000, // 1 megabit
            audioChannels: 2,
            audioSampleRate: 44100,
            audioBitrate: 128000, // 128 kilobits
            progress: function(info) {
                console.log('transcodeVideo progress callback, info: ' + info);
            }
        }
    );
}

function videoTranscodeSuccess(result) {
	// result is the path to the transcoded video on the device
    console.log('videoTranscodeSuccess, result: ' + result);
}

function videoTranscodeError(err) {
	console.log('videoTranscodeError, err: ' + err);
}
```

#### Windows Quirks
Windows does not support any of the optional parameters at this time. Specifying them will not cause an error but, there is no functionality behind them.

### Trim a Video (iOS only)
```javascript
VideoEditor.trim(
    trimSuccess,
    trimFail,
    {
        fileUri: 'file-uri-here', // path to input video
        trimStart: 5, // time to start trimming in seconds
        trimEnd: 15, // time to end trimming in seconds
        outputFileName: 'output-name', // output file name
        progress: function(info) {} // optional, see docs on progress
    }
);

function trimSuccess(result) {
    // result is the path to the trimmed video on the device
    console.log('trimSuccess, result: ' + result);
}

function trimFail(err) {
    console.log('trimFail, err: ' + err);
}
```

### Create a JPEG thumbnail from a video
```javascript
VideoEditor.createThumbnail(
    success, // success cb
    error, // error cb
    {
        fileUri: 'file-uri-here', // the path to the video on the device
        outputFileName: 'output-name', // the file name for the JPEG image
        atTime: 2, // optional, location in the video to create the thumbnail (in seconds)
        width: 320, // optional, width of the thumbnail
        height: 480, // optional, height of the thumbnail
        quality: 100 // optional, quality of the thumbnail (between 1 and 100)
    }
);
// atTime will default to 0 if not provided
// width and height will be the same as the video input if they are not provided
// quality will default to 100 if not provided
```

```javascript
// this example uses the cordova media capture plugin
navigator.device.capture.captureVideo(
    videoCaptureSuccess,
    videoCaptureError,
    {
        limit: 1,
        duration: 20
    }
);

function videoCaptureSuccess(mediaFiles) {
    // Wrap this below in a ~100 ms timeout on Android if
    // you just recorded the video using the capture plugin.
    // For some reason it is not available immediately in the file system.

    var file = mediaFiles[0];
    var videoFileName = 'video-name-here'; // I suggest a uuid

    VideoEditor.createThumbnail(
        createThumbnailSuccess,
        createThumbnailError,
        {
            fileUri: file.fullPath,
            outputFileName: videoFileName,
            atTime: 2,
            width: 320,
            height: 480,
            quality: 100
        }
    );
}

function createThumbnailSuccess(result) {
    // result is the path to the jpeg image on the device
    console.log('createThumbnailSuccess, result: ' + result);
}
```

#### A note on width and height used by createThumbnail
The aspect ratio of the thumbnail created will match that of the video input.  This means you may not get exactly the width and height dimensions you give to `createThumbnail` for the jpeg.  This for your convenience but let us know if it is a problem.  I am considering adding a `maintainAspectRatio` option to `createThumbnail` (and when this option is false you might have stretched, square thumbnails :laughing:).

### Get info on a video (width, height, orientation, duration, size, & bitrate)
```javascript
VideoEditor.getVideoInfo(
    success, // success cb
    error, // error cb
    {
        fileUri: 'file-uri-here', // the path to the video on the device
    }
);
```

```javascript
VideoEditor.getVideoInfo(
    getVideoInfoSuccess,
    getVideoInfoError,
    {
        fileUri: file.fullPath
    }
);

function getVideoInfoSuccess(info) {
    console.log('getVideoInfoSuccess, info: ' + JSON.stringify(info, null, 2));
    // info is a JSON object with the following properties -
    {
        width: 1920,
        height: 1080,
        orientation: 'landscape', // will be portrait or landscape
        duration: 3.541, // duration in seconds
        size: 6830126, // size of the video in bytes
        bitrate: 15429777 // bitrate of the video in bits per second,
        videoMediaType: 'video/3gpp' // Media type of the video, android example: 'video/3gpp', ios example: 'avc1',
        audioMediaType: 'audio/mp4a-latm' // Media type of the audio track in video, android example: 'audio/mp4a-latm', ios example: 'aac',
    }
}
```

## Android & FFmpeg
FFmpeg has been removed from android for several reasons but mainly for performance.  If you still need the old functionality that FFmpeg provided  [V1.09](https://github.com/okanbeydanol/cordova-plugin-video-editor/tree/1.0.9) is the last version that will use it.

## On iOS

[iOS Developer AVFoundation Documentation](https://developer.apple.com/library/ios/documentation/AudioVideo/Conceptual/AVFoundationPG/Articles/01_UsingAssets.html#//apple_ref/doc/uid/TP40010188-CH7-SW8)

[Video compression in AVFoundation](http://www.iphonedevsdk.com/forum/iphone-sdk-development/110246-video-compression-avassetwriter-in-avfoundation.html)

[AVFoundation slides - tips/tricks](https://speakerdeck.com/bobmccune/composing-and-editing-media-with-av-foundation)

[AVFoundation slides #2](http://www.slideshare.net/bobmccune/learning-avfoundation)

[Bob McCune's AVFoundation Editor - ios app example](https://github.com/tapharmonic/AVFoundationEditor)

[Saving videos after recording videos](http://stackoverflow.com/questions/20902234/save-video-to-library-after-capturing-video-using-phonegap-capturevideo)



## On Android

[Android Documentation](http://developer.android.com/guide/appendix/media-formats.html#recommendations)

[Android Media Stores](http://developer.android.com/reference/android/provider/MediaStore.html#EXTRA_VIDEO_QUALITY)

[How to Port ffmpeg (the Program) to Android–Ideas and Thoughts](http://www.roman10.net/how-to-port-ffmpeg-the-program-to-androidideas-and-thoughts/)

[How to Build Android Applications Based on FFmpeg by An Example](http://www.roman10.net/how-to-build-android-applications-based-on-ffmpeg-by-an-example/)


## On Windows


## License

Android: Apache 2.0

iOS: MIT

Windows: Apache 2.0


CHANGELOG.md
# Change Log

## [1.0.0](https://github.com/okanbeydanol/cordova-plugin-video-editor/tree/1.0.0) (2015-11-11)
[Full Changelog](https://github.com/okanbeydanol/cordova-plugin-video-editor/compare/0.0.6...1.0.0)

**Implemented enhancements:**

- trim start point [\#16](https://github.com/okanbeydanol/cordova-plugin-video-editor/issues/16)

**Closed issues:**

- VideoEditor reference error [\#31](https://github.com/okanbeydanol/cordova-plugin-video-editor/issues/31)
- Setting trim start [\#28](https://github.com/okanbeydanol/cordova-plugin-video-editor/issues/28)
- Stretched and rotated video when in portrait mode [\#9](https://github.com/okanbeydanol/cordova-plugin-video-editor/issues/9)

## [0.0.6](https://github.com/okanbeydanol/cordova-plugin-video-editor/tree/0.0.6) (2015-11-11)
[Full Changelog](https://github.com/okanbeydanol/cordova-plugin-video-editor/compare/0.0.5...0.0.6)

## [0.0.5](https://github.com/okanbeydanol/cordova-plugin-video-editor/tree/0.0.5) (2015-11-10)
[Full Changelog](https://github.com/okanbeydanol/cordova-plugin-video-editor/compare/0.0.4...0.0.5)

**Closed issues:**

- Selecting Videos from Gallery for transcoding [\#33](https://github.com/okanbeydanol/cordova-plugin-video-editor/issues/33)
- is trimming working any android version? [\#27](https://github.com/okanbeydanol/cordova-plugin-video-editor/issues/27)
- Input files are remove by default after encoding [\#23](https://github.com/okanbeydanol/cordova-plugin-video-editor/issues/23)
- Example project installation or installation instructions? [\#22](https://github.com/okanbeydanol/cordova-plugin-video-editor/issues/22)
- Transcoding not working in Jelly Bean [\#15](https://github.com/okanbeydanol/cordova-plugin-video-editor/issues/15)
- Not valid import com.netcompss.loader.LoadJNI; [\#8](https://github.com/okanbeydanol/cordova-plugin-video-editor/issues/8)

## [0.0.4](https://github.com/okanbeydanol/cordova-plugin-video-editor/tree/0.0.4) (2015-11-09)
**Closed issues:**

- Plugin won't compile with Cordova 5 [\#26](https://github.com/okanbeydanol/cordova-plugin-video-editor/issues/26)
- Error generating manifest, icon value also present in... [\#24](https://github.com/okanbeydanol/cordova-plugin-video-editor/issues/24)
- Consider switching from ffmpeg4android library as it costs $400 [\#20](https://github.com/okanbeydanol/cordova-plugin-video-editor/issues/20)
- File not Found but ... file exists [\#11](https://github.com/okanbeydanol/cordova-plugin-video-editor/issues/11)
- Any Documentation? [\#1](https://github.com/okanbeydanol/cordova-plugin-video-editor/issues/1)

**Merged pull requests:**

- Abandon ffmpeg4android\_lib in favor of android-ffmpeg-java [\#13](https://github.com/okanbeydanol/cordova-plugin-video-editor/pull/13) ([rossmartin](https://github.com/rossmartin))
- Add option to save to gallery \(android\). [\#12](https://github.com/okanbeydanol/cordova-plugin-video-editor/pull/12) ([rossmartin](https://github.com/rossmartin))
- Add createThumbnail support [\#5](https://github.com/okanbeydanol/cordova-plugin-video-editor/pull/5) ([rossmartin](https://github.com/rossmartin))
- Updated readme \(VideoEditorPlugin --\> VideoEditor\) [\#4](https://github.com/okanbeydanol/cordova-plugin-video-editor/pull/4) ([rossmartin](https://github.com/rossmartin))
- Added transcoding support for android and video trimming. [\#3](https://github.com/okanbeydanol/cordova-plugin-video-editor/pull/3) ([rossmartin](https://github.com/rossmartin))
- Fixed JS errors in plugin files and updated transcoding [\#2](https://github.com/okanbeydanol/cordova-plugin-video-editor/pull/2) ([rossmartin](https://github.com/rossmartin))



\* *This Change Log was automatically generated by [github_changelog_generator](https://github.com/skywinder/Github-Changelog-Generator)*

src/android/VideoEditor.java
package org.apache.cordova.videoeditor;

import java.io.*;
import java.text.SimpleDateFormat;
import java.util.Date;
import java.util.Locale;

import android.graphics.Bitmap;

import org.apache.cordova.CordovaInterface;
import org.apache.cordova.CordovaPlugin;
import org.apache.cordova.CallbackContext;
import org.apache.cordova.CordovaResourceApi;
import org.apache.cordova.CordovaWebView;
import org.apache.cordova.PluginResult;
import org.json.JSONArray;
import org.json.JSONException;
import org.json.JSONObject;

import android.content.Context;
import android.content.Intent;
import android.content.pm.ApplicationInfo;
import android.content.pm.PackageManager;
import android.content.pm.PackageManager.NameNotFoundException;
import android.media.MediaExtractor;
import android.media.MediaMetadataRetriever;
import android.net.Uri;
import android.os.Environment;

import net.ypresto.androidtranscoder.MediaTranscoder;
import net.ypresto.androidtranscoder.utils.MediaExtractorUtils;

/**
 * VideoEditor plugin for Android
 * Created by Ross Martin 2-2-15
 */
public class VideoEditor extends CordovaPlugin {

    private static final String TAG = "VideoEditor";

    private CallbackContext callback;
    private CordovaResourceApi resourceApi;

    /**
     * Initialization
     */
    @Override
    public void initialize(CordovaInterface cordova, CordovaWebView webView) {
        super.initialize(cordova, webView);
        this.resourceApi = webView.getResourceApi();
    }

    /**
     * Executes the request to the plugin.
     */
    @Override
    public boolean execute(String action, JSONArray args, CallbackContext callbackContext) throws JSONException {
        this.callback = callbackContext;
        switch (action) {
            case "transcodeVideo":
                try {
                    this.transcodeVideo(args);
                } catch (IOException e) {
                    callback.error(e.toString());
                }
                return true;
            case "createThumbnail":
                try {
                    this.createThumbnail(args);
                } catch (IOException e) {
                    callback.error(e.toString());
                }
                return true;
            case "getVideoInfo":
                try {
                    this.getVideoInfo(args);
                } catch (IOException e) {
                    callback.error(e.toString());
                }
                return true;
        }
        return false;
    }

    /**
     * transcodeVideo
     *
     * Transcodes a video
     *
     * ARGUMENTS
     * =========
     *
     * fileUri                     - path to input video
     * outputFileName              - output file name
     * saveToLibrary               - save to gallery
     * width                       - width for the output video
     * height                      - height for the output video
     * fps                         - fps the video
     * videoBitrate                - video bitrate for the output video in bits
     * audioBitrate                - audio bitrate for the output video in bits
     * audioChannels               - number of audio channels
     * skipVideoTranscodingIfAVC   - skip any transcoding actions (conversion/resizing/etc..) if the input video is avc video
     * duration                    - max video duration (in seconds?)
     *
     * RESPONSE
     * ========
     *
     * outputFilePath - path to output file
     *
     * @param args arguments
     */
    private void transcodeVideo(JSONArray args) throws JSONException, IOException {
        JSONObject options = args.optJSONObject(0);
        final ReadDataResult readResult = this.readDataFrom(options.getString("fileUri"));
        if (readResult == null) {
            return;
        }

        final String outputFileName = options.optString(
                "outputFileName",
                new SimpleDateFormat("yyyyMMdd_HHmmss", Locale.ENGLISH).format(new Date())
        );

        final int width = options.optInt("width", CustomAndroidFormatStrategy.DEFAULT_WIDTH);
        final int height = options.optInt("height", CustomAndroidFormatStrategy.DEFAULT_HEIGHT);
        final int fps = options.optInt("fps", CustomAndroidFormatStrategy.DEFAULT_FRAMERATE);
        final int videoBitrate = options.optInt("videoBitrate", CustomAndroidFormatStrategy.DEFAULT_VIDEO_BITRATE); // default to 9 megabit
        final int audioBitrate = options.optInt("audioBitrate", CustomAndroidFormatStrategy.AUDIO_BITRATE_AS_IS);
        final int audioChannels = options.optInt("audioChannels", CustomAndroidFormatStrategy.AUDIO_CHANNELS_AS_IS);
        final boolean skipVideoTranscodingIfAVC = options.optBoolean("skipVideoTranscodingIfAVC", CustomAndroidFormatStrategy.DEFAULT_SKIP_AVC_VIDEO_TRANSCODING);
        final long videoDuration = options.optLong("duration", 0) * 1000 * 1000;
        final String outputExtension = ".mp4";

        final Context appContext = cordova.getActivity().getApplicationContext();
        final PackageManager pm = appContext.getPackageManager();

        ApplicationInfo ai;
        try {
            ai = pm.getApplicationInfo(cordova.getActivity().getPackageName(), 0);
        } catch (final NameNotFoundException e) {
            ai = null;
        }
        final String appName = (String) (ai != null ? pm.getApplicationLabel(ai) : "Unknown");

        final boolean saveToLibrary = options.optBoolean("saveToLibrary", true);
        File mediaStorageDir;

        if (saveToLibrary) {
            mediaStorageDir = new File(
                    Environment.getExternalStorageDirectory() + "/Movies",
                    appName
            );
        } else {
            mediaStorageDir = new File(Environment.getExternalStorageDirectory().getAbsolutePath() + "/Android/data/" + cordova.getActivity().getPackageName() + "/files/files/videos");
        }

        if (!mediaStorageDir.exists()) {
            if (!mediaStorageDir.mkdirs()) {
                callback.error("Can't access or make Movies directory");
                readResult.close();
                return;
            }
        }

        final String outputFilePath = new File(
                mediaStorageDir.getPath(),
                outputFileName + outputExtension
        ).getAbsolutePath();

        cordova.getThreadPool().execute(() -> {

            try {
                MediaTranscoder.Listener listener = new MediaTranscoder.Listener() {
                    @Override
                    public void onTranscodeProgress(double progress) {
                        JSONObject jsonObj = new JSONObject();
                        try {
                            jsonObj.put("progress", progress);
                        } catch (JSONException e) {
                            e.printStackTrace();
                        }

                        PluginResult progressResult = new PluginResult(PluginResult.Status.OK, jsonObj);
                        progressResult.setKeepCallback(true);
                        callback.sendPluginResult(progressResult);
                    }

                    @Override
                    public void onTranscodeCompleted() {
                        File outFile = new File(outputFilePath);
                        if (!outFile.exists()) {
                            readResult.close();
                            callback.error("an error ocurred during transcoding");
                            return;
                        }

                        // make the gallery display the new file if saving to library
                        if (saveToLibrary) {
                            Intent scanIntent = new Intent(Intent.ACTION_MEDIA_SCANNER_SCAN_FILE);
                            scanIntent.setData(readResult.result.uri);
                            scanIntent.setData(Uri.fromFile(outFile));
                            appContext.sendBroadcast(scanIntent);
                        }

                        readResult.close();
                        callback.success(outputFilePath);
                    }

                    @Override
                    public void onTranscodeCanceled() {
                        readResult.close();
                        callback.error("transcode canceled");
                    }

                    @Override
                    public void onTranscodeFailed(Exception exception) {
                        readResult.close();
                        callback.error(exception.toString());
                    }
                };

                final FileDescriptor fileDescriptor = readResult.getFD();

                MediaMetadataRetriever mmr = new MediaMetadataRetriever();
                mmr.setDataSource(fileDescriptor);

                MediaTranscoder.getInstance().transcodeVideo(
                        fileDescriptor,
                        outputFilePath,
                        new CustomAndroidFormatStrategy(videoBitrate, fps, width, height, audioBitrate, audioChannels, skipVideoTranscodingIfAVC),
                        listener,
                        videoDuration
                );

            } catch (Throwable e) {
                readResult.close();
                callback.error(e.toString());
            }

        });
    }

    /**
     * createThumbnail
     *
     * Creates a thumbnail from the start of a video.
     *
     * ARGUMENTS
     * =========
     * fileUri        - input file path
     * outputFileName - output file name
     * atTime         - location in the video to create the thumbnail (in seconds)
     * width          - width for the thumbnail (optional)
     * height         - height for the thumbnail (optional)
     * quality        - quality of the thumbnail (optional, between 1 and 100)
     *
     * RESPONSE
     * ========
     *
     * outputFilePath - path to output file
     *
     * @param args arguments
     */
    private void createThumbnail(JSONArray args) throws JSONException, IOException {
        JSONObject options = args.optJSONObject(0);

        final ReadDataResult readResult = this.readDataFrom(options.getString("fileUri"));
        if (readResult == null) {
            return;
        }

        String outputFileName = options.optString(
                "outputFileName",
                new SimpleDateFormat("yyyyMMdd_HHmmss", Locale.ENGLISH).format(new Date())
        );

        final int quality = options.optInt("quality", 100);
        final int width = options.optInt("width", 0);
        final int height = options.optInt("height", 0);
        long atTimeOpt = options.optLong("atTime", 0);
        final long atTime = (atTimeOpt == 0) ? 0 : atTimeOpt * 1000000;

        File externalFilesDir =  new File(Environment.getExternalStorageDirectory().getAbsolutePath() + "/Android/data/" + cordova.getActivity().getPackageName() + "/files/files/videos");
        if (!externalFilesDir.exists()) {
            if (!externalFilesDir.mkdirs()) {
                callback.error("Can't access or make Movies directory");
                readResult.close();
                return;
            }
        }

        final File outputFile =  new File(
                externalFilesDir.getPath(),
                outputFileName + ".jpg"
        );
        final String outputFilePath = outputFile.getAbsolutePath();

        // start task
        cordova.getThreadPool().execute(() -> {

            OutputStream outStream = null;

            try {
                final FileDescriptor fileDescriptor = readResult.getFD();
                MediaMetadataRetriever mmr = new MediaMetadataRetriever();
                mmr.setDataSource(fileDescriptor);

                Bitmap bitmap = mmr.getFrameAtTime(atTime);

                if (width > 0 || height > 0) {
                    int videoWidth = bitmap.getWidth();
                    int videoHeight = bitmap.getHeight();
                    double aspectRatio = (double) videoWidth / (double) videoHeight;

                    int scaleWidth = Double.valueOf(height * aspectRatio).intValue();
                    int scaleHeight = Double.valueOf(scaleWidth / aspectRatio).intValue();

                    final Bitmap resizedBitmap = Bitmap.createScaledBitmap(bitmap, scaleWidth, scaleHeight, false);
                    bitmap.recycle();
                    bitmap = resizedBitmap;
                }

                outStream = new FileOutputStream(outputFile);
                bitmap.compress(Bitmap.CompressFormat.JPEG, quality, outStream);

                callback.success(outputFilePath);

            } catch (Throwable e) {
                if (outStream != null) {
                    try {
                        outStream.close();
                    } catch (IOException e1) {
                        e1.printStackTrace();
                    }
                }

                callback.error(e.toString());

            } finally {
                readResult.close();
            }

        });
    }

    /**
     * getVideoInfo
     *
     * Gets info on a video
     *
     * ARGUMENTS
     * =========
     *
     * fileUri:      - path to input video
     *
     * RESPONSE
     * ========
     *
     * width              - width of the video
     * height             - height of the video
     * orientation        - orientation of the video
     * duration           - duration of the video (in seconds)
     * size               - size of the video (in bytes)
     * bitrate            - bitrate of the video (in bits per second)
     * videoMediaType     - Media type of the video
     * audioMediaType     - Media type of the audio track in video
     *
     * @param args arguments
     */
    private void getVideoInfo(JSONArray args) throws JSONException, IOException {
        JSONObject options = args.optJSONObject(0);

        final ReadDataResult readResult = this.readDataFrom(options.getString("fileUri"));
        if (readResult == null) {
            return;
        }

        final FileDescriptor fileDescriptor = readResult.getFD();
        MediaMetadataRetriever mmr = new MediaMetadataRetriever();
        mmr.setDataSource(fileDescriptor);
        float videoWidth = Float.parseFloat(mmr.extractMetadata(MediaMetadataRetriever.METADATA_KEY_VIDEO_WIDTH));
        float videoHeight = Float.parseFloat(mmr.extractMetadata(MediaMetadataRetriever.METADATA_KEY_VIDEO_HEIGHT));

        String orientation;
        String mmrOrientation = mmr.extractMetadata(MediaMetadataRetriever.METADATA_KEY_VIDEO_ROTATION);

        if (videoWidth < videoHeight) {
            if (mmrOrientation.equals("0") || mmrOrientation.equals("180")) {
                orientation = "portrait";
            } else {
                orientation = "landscape";
            }
        } else {
            if (mmrOrientation.equals("0") || mmrOrientation.equals("180")) {
                orientation = "landscape";
            } else {
                orientation = "portrait";
            }
        }

        double duration = Double.parseDouble(mmr.extractMetadata(MediaMetadataRetriever.METADATA_KEY_DURATION)) / 1000.0;
        long bitrate = Long.parseLong(mmr.extractMetadata(MediaMetadataRetriever.METADATA_KEY_BITRATE));

        String videoMediaType;
        String audioMediaType;
        try {
            final MediaExtractor mExtractor = new MediaExtractor();
            mExtractor.setDataSource(fileDescriptor);
            MediaExtractorUtils.TrackResult trackResult = MediaExtractorUtils.getFirstVideoAndAudioTrack(mExtractor);

            // get types
            videoMediaType = trackResult.mVideoTrackMime;
            audioMediaType = trackResult.mAudioTrackMime;

            // release resources
            mExtractor.release();
            trackResult = null;
        } catch (Throwable e) {
            callback.error(e.toString());
            readResult.close();
            return;
        }

        JSONObject response = new JSONObject();
        response.put("width", videoWidth);
        response.put("height", videoHeight);
        response.put("orientation", orientation);
        response.put("duration", duration);
        response.put("size", readResult.result.length);
        response.put("bitrate", bitrate);
        response.put("videoMediaType", videoMediaType);
        response.put("audioMediaType", audioMediaType);

        // release resources
        readResult.close();

        callback.success(response);
    }

    /**
     * Reads the data by the given url
     * @param url the url to read the data
     * @return results of the reading
     */
    private ReadDataResult readDataFrom(String url) throws IOException {
        if (!FileUtils.isLocal(url)) {
            final String msg = "The provided url is null or not local: " + url;
            callback.error(msg);
            return null;
        }

        final Context context = this.cordova.getActivity().getApplicationContext();
        Uri uri = Uri.parse(url);
        if (uri.isRelative()) {
            uri = Uri.parse(FileUtils.getPath(context, uri));
        }

        CordovaResourceApi.OpenForReadResult readResult = resourceApi.openForRead(uri, true);
        return new ReadDataResult(readResult);
    }

    /**
     * Simple wrapper over the CordovaResourceApi.OpenForReadResult with some util methods
     */
    public static final class ReadDataResult {
        public final CordovaResourceApi.OpenForReadResult result;
        private FileDescriptor fileDescriptor;

        public ReadDataResult(CordovaResourceApi.OpenForReadResult result) {
            this.result = result;
        }

        /**
         * Returns file descriptor based on the OpenForReadResult
         * @return FileDescriptor for the given result
         */
        public FileDescriptor getFD() throws IOException {
            if (this.fileDescriptor != null) {
                return this.fileDescriptor;
            }

            if (this.result.inputStream != null &&
                    this.result.inputStream instanceof FileInputStream) {
                this.fileDescriptor = ((FileInputStream) this.result.inputStream).getFD();
                return this.fileDescriptor;
            }

            if (this.result.assetFd != null) {
                this.fileDescriptor = this.result.assetFd.getFileDescriptor();
                return this.fileDescriptor;
            }

            return null;
        }

        /**
         * Closes the stream and descriptor if them exists
         */
        public void close() {
            try {
                if (this.result.assetFd != null) {
                    this.result.assetFd.close();
                }
                if (this.result.inputStream != null) {
                    this.result.inputStream.close();
                }
            } catch (IOException e) {
                e.printStackTrace();
            }
        }
    }

}
